Mathlib/RingTheory/GradedAlgebra/Homogeneous/Submodule.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+import Mathlib.RingTheory.GradedAlgebra.Basic
+import Mathlib.Algebra.GradedMulAction
+
+/-!
+# Homogeneous submodules of a graded module
+
+This file defines homogeneous submodule of a graded module `⨁ᵢ ℳᵢ` over graded ring `⨁ᵢ 𝒜ᵢ` and
+operations on them.
+
+## Main definitions
+
+For any `p : Submodule A M`:
+* `Submodule.IsHomogeneous ℳ p`: The property that a submodule is closed under `GradedModule.proj`.
+* `HomogeneousSubmodule 𝒜 ℳ`: The structure extending submodules which satisfy
+  `Submodule.IsHomogeneous`.
+
+## Implementation notes
+
+The **notion** of homogeneous submodule does not rely on a graded ring, only a decomposition of the
+the module. However, most interesting properties of homogeneous submodules do rely on the base ring
+being a graded ring. For technical reasons, we make `HomogeneousSubmodule` depend on a graded ring.
+For example, if the definition of a homogeneous submodule does not depend on a graded ring, the
+instance that `HomogeneousSubmodule` is a complete lattice cannot be synthesized due to
+synthesation order.
+
+## Tags
+
+graded algebra, homogeneous
+-/
+
+open SetLike DirectSum Pointwise Set
+
+variable {ιA ιM σA σM A M : Type*}
+
+variable [Semiring A] [AddCommMonoid M] [Module A M]
+
+section HomogeneousDef
+
+/--
+An `A`-submodule `p ⊆ M` is homogeneous if for every `m ∈ p`, all homogeneous components of `m` are
+in `p`.
+-/
+def Submodule.IsHomogeneous (p : Submodule A M) (ℳ : ιM → σM)
+    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ] : Prop :=
+  SetLike.IsHomogeneous ℳ p
+
+theorem Submodule.IsHomogeneous.mem_iff {p : Submodule A M}
+    (ℳ : ιM → σM)
+    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ]
+    (hp : p.IsHomogeneous ℳ) {x} :
+    x ∈ p ↔ ∀ i, (decompose ℳ x i : M) ∈ p :=
+  AddSubmonoidClass.IsHomogeneous.mem_iff ℳ _ hp
+
+/-- For any `Semiring A`, we collect the homogeneous submodule of `A`-modules into a type. -/
+structure HomogeneousSubmodule (𝒜 : ιA → σA) (ℳ : ιM → σM)
+    [DecidableEq ιA] [AddMonoid ιA] [SetLike σA A] [AddSubmonoidClass σA A] [GradedRing 𝒜]
+    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ]
+    [VAdd ιA ιM] [GradedSMul 𝒜 ℳ]
+    extends Submodule A M where
+  is_homogeneous' : toSubmodule.IsHomogeneous ℳ
```

Comment:
Do we want `AddSubmonoidClass` and `SMulMemClass` on this?

Reply:
Yes, these two are required in `DirectSum.Decomposition`

Reply:
I meant `SMulMemClass (HomogeneousSubmodule 𝒜 ℳ) A M`, which I don't think is present in the current PR?

Reply:
done